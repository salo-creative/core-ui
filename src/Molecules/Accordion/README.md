# Accordion

This component functions as an accordion component like you might find anywhere on the web. All direct children of the Accordion are displayed inside Accordion Sections. They must have title, and they can optionally take an iconBefore argument that determines whether or not to display an icon before the title.

The accordion does not return values by itself, but instead just dresses up and organises other components.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Accordion, { AccordionItem } from '@salo/core-ui/Molecules/Accordion';
```

Implement as follows

```javascript
<Accordion>
  <AccordionItem
    id='1'
    label='label 1'
  >
   // Your content
  </AccordionItem>
  <AccordionItem
    id='2'
    label='label 2'
  >
   // Your content
  </AccordionItem>
</Accordion>
```