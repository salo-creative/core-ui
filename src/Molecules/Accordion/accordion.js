import React from 'react';
import PropTypes from 'prop-types';
import { indexOf } from 'lodash';
// COMPONENTS & STYLES

const Accordion = (props) => {
  const {
    allowMultiple,
    children,
    initialOpen
  } = props;

  const [active, setActive] = React.useState(initialOpen);

  const handleClick = (id) => {
    const i = indexOf(active, id);
    if (i >= 0) { // remove active state for this id
      setActive([
        ...active.slice(0, i),
        ...active.slice(i + 1)
      ]);
    } else { // Make active and handle multiple bool
      setActive(allowMultiple ? [...active, id] : [id]);
    }
  };

  const items = children.map(item => {
    const { id } = item.props;
    return {
      ...item,
      props: {
        ...item.props,
        onClick: () => handleClick(id),
        isOpen: !!active.includes(id)
      }
    };
  });
  return items;
};

Accordion.defaultProps = {
  allowMultiple: false,
  initialOpen: []
};

Accordion.propTypes = {
  allowMultiple: PropTypes.bool,
  children: PropTypes.any.isRequired,
  initialOpen: PropTypes.arrayOf(PropTypes.oneOf([PropTypes.string, PropTypes.number]))
};

export default Accordion;