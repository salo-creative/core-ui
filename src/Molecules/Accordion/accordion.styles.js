import styled from 'styled-components';
import Icon from '@salo/icons';

export const Wrapper = styled.div`
  width: 100%;
`;

export const Header = styled.button`
  outline: none;
  border: none;
  display: flex;
  align-items: center;
  background-color: ${ ({ theme }) => theme.paleGrey };
  transition: background-color 0.2s linear;
  border-bottom: 1px solid ${ ({ theme }) => theme.grey };
  cursor: pointer;
  height: 5rem;
  width: 100%;
  text-align: left;
  padding: 0 1rem;
`;

export const Body = styled.div`
  display: inline-block;
  width: 100%;
  padding: 1rem;
`;

export const Label = styled.p`
  margin: 0 1rem;
  font-size: 1.4rem;
  width: 100%;
`;

export const Chevron = styled.span`
  display: flex;
  transition: transform 0.3s linear;
  transform: ${ ({ isOpen }) => (isOpen ? 'rotate(180deg)' : 'rotate(0deg)') };
`;