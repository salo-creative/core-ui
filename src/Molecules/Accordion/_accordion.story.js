import React from 'react';
import { withKnobs, boolean, text } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// Tests.
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// Featured Component.
import { Accordion, AccordionItem } from '../../index';
import P from '../../Typography/P';

// README
import README from './README.md';

export const Basic = () => {
  // Store
  const store = new Store({});
  // Knobs.
  const iconBefore = text('iconBefore', 'pencil');
  const allowMultiple = boolean('allowMultiple', false);
  return (
    <RenderWithProps store={ store }>
      <Accordion allowMultiple={ allowMultiple }>
        <AccordionItem label='Interesting Label' iconBefore={ iconBefore } id='1'>
          <P>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis porttitor vestibulum mi id
            blandit. In pulvinar varius mauris, eget faucibus dui molestie ut. Donec at neque in
            diam mollis laoreet. Curabitur lobortis, augue vel maximus dictum, leo neque gravida
            dolor, sit amet ullamcorper magna velit sit amet magna. Duis ac dolor lacus. Aenean sit
            amet nisi nec eros scelerisque rutrum. Fusce mattis a erat at suscipit. Suspendisse
            malesuada ultrices lorem, quis bibendum massa tempor et. Ut interdum tellus quis diam
            venenatis, eget faucibus libero blandit.
          </P>
          <P>
            Phasellus interdum tempus odio, vel vestibulum diam accumsan at. Nulla facilisi. Ut
            justo lectus, luctus a vulputate vel, consectetur nec ante. Ut auctor ut metus sed
            semper. In elit ex, dictum ac tempor quis, feugiat et purus. Vestibulum non lorem ut
            lacus sagittis interdum quis sit amet ipsum. Praesent a lectus ultrices ante aliquet
            pulvinar. Sed odio erat, varius ut finibus eu, molestie ac sem. Aliquam eu mollis dolor.
            Suspendisse nec feugiat ligula. Aliquam hendrerit sem quam, at hendrerit elit viverra
            ut. Maecenas porttitor turpis eget augue tristique faucibus. Proin non erat viverra,
            vestibulum libero eget, volutpat leo. Praesent vestibulum pulvinar mi, at pulvinar eros
            tempor quis. Cras a semper urna.
          </P>
        </AccordionItem>
        <AccordionItem label='A Longer Label, Just As Interesting' iconBefore={ iconBefore } id='2'>
          <P>
            Curabitur ultrices est a consequat aliquet. Donec varius finibus augue eget facilisis.
            Mauris ut mauris ullamcorper, tempor eros quis, dapibus orci. Donec dolor elit, cursus
            in tempor et, consectetur ut augue. Nunc sodales vitae nunc ac suscipit. Curabitur vitae
            quam sit amet tortor dictum lacinia. Cras eleifend urna quis efficitur viverra.
            Curabitur vel consequat odio, quis elementum metus. Fusce ac molestie est. Nam pharetra
            ligula at neque mattis, et tristique lorem sagittis. Curabitur a molestie justo.
          </P>
        </AccordionItem>
        <AccordionItem id='3' label='A Third Label' iconBefore={ iconBefore }>
          <P>
            Sed eu sollicitudin metus. Integer nulla massa, porttitor vel vulputate vitae, pulvinar
            ac enim. Donec eu rhoncus turpis. Phasellus venenatis ante neque, at tristique lacus
            elementum eget. Vivamus aliquam, tellus eget venenatis volutpat, lorem leo hendrerit
            tellus, id scelerisque tellus libero id ante. Suspendisse potenti. Aliquam ex ante,
            efficitur sit amet neque in, consectetur accumsan magna. Proin tincidunt feugiat sapien
            vitae aliquam. Aliquam sed libero ut ligula facilisis vestibulum ac ac odio. Ut ac
            commodo enim. Sed vitae ligula ut sapien vehicula placerat eget pretium arcu. Mauris
            quis enim mauris. Pellentesque dolor ante, tempor ac laoreet dapibus, aliquam at orci.
            Aliquam mauris velit, luctus in massa vel, fringilla volutpat leo.
          </P>
        </AccordionItem>
      </Accordion>
    </RenderWithProps>
  );
};

Basic.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    jest: ['accordion'],
    info: {
      propTablesExclude: [RenderWithProps, P]
    },
    notes: README
  }
};

export default {
  title: 'Molecules/Accordion'
};