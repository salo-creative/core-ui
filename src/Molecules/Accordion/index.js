import Accordion from './accordion';

export { default as AccordionItem } from './accordion.item';

export default Accordion;