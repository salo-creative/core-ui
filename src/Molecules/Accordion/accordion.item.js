import React from 'react';
import PropTypes from 'prop-types';
import Icon from '@salo/icons';

// COMPONENTS & STYLES
import { Wrapper, Header, Body, Label, Chevron } from './accordion.styles';

const AccordionItem = (props) => {
  const {
    children,
    iconBefore,
    isOpen,
    label,
    onClick
  } = props;
  
  return (
    <Wrapper>
      <Header
        onClick={ onClick }
      >
        { iconBefore && <Icon size={ 20 } icon={ iconBefore } /> }
        <Label>{ label }</Label>
        <Chevron isOpen={ isOpen }>
          <Icon
            icon='chevron_down'
            isOpen={ isOpen }
            size={ 20 }
          />
        </Chevron>
      </Header>
      { isOpen && (
        <Body>
          { children }
        </Body>
      ) }
    </Wrapper>
  );
};

AccordionItem.defaultProps = {
  iconBefore: '',
  isOpen: false,
  label: 'Accordion item',
  onClick: () => null
};

AccordionItem.propTypes = {
  children: PropTypes.any.isRequired,
  isOpen: PropTypes.bool,
  iconBefore: PropTypes.string,
  label: PropTypes.string,
  onClick: PropTypes.func
};

export default AccordionItem;