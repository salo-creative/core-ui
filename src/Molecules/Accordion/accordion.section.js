import React from 'react';
import PropTypes from 'prop-types';

// Components.
import Icon from '@salo/icons';
import { SectionHead, SectionBody } from './accordion.styles';
import H3 from '../../Typography/H3';

const AccordionSection = (props) => {
  const {
    children,
    isOpen,
    title,
    onClick,
    iconBefore
  } = props;

  const sectionClick = () => {
    onClick(title);
  };

  return (
    <div>
      <SectionHead onClick={ sectionClick } isOpen={ isOpen }>
        { iconBefore && (
          <Icon icon={ iconBefore } margin='0 0 0 1rem' />
        ) }
        <H3 margin='1rem'>
          { title }
        </H3>
        { isOpen && (
          <Icon icon='chevron_down' margin='0 1rem 0 0' size={ 35 } />
        ) }
        { !isOpen && (
          <Icon icon='chevron_up' margin='0 1rem 0 0' size={ 35 } />
        ) }
      </SectionHead>
      { isOpen && (
        <SectionBody>
          { children }
        </SectionBody>
      ) }
    </div>
  );
};

AccordionSection.defaultProps = {
  iconBefore: null,
  isOpen: false
};

AccordionSection.propTypes = {
  children: PropTypes.instanceOf(Object).isRequired,
  isOpen: PropTypes.bool,
  title: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  iconBefore: PropTypes.string
};

export default AccordionSection;