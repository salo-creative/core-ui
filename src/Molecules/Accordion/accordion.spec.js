import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Accordion from './index';
import AccordionItem from './accordion.item';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Accordion>
      <AccordionItem
        id='1'
        label='label 1'
      >
        Item 1
      </AccordionItem>
      <AccordionItem
        id='2'
        label='label 2'
      >
        Item 2
      </AccordionItem>
    </Accordion>
  );
  expect(container.firstChild).toMatchSnapshot();
});