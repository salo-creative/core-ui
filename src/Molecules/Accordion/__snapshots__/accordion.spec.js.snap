// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c4 {
  height: 20px;
  width: 20px;
  margin: 0;
}

.c0 {
  width: 100%;
}

.c1 {
  outline: none;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #fafcff;
  -webkit-transition: background-color 0.2s linear;
  transition: background-color 0.2s linear;
  border-bottom: 1px solid #e9ecef;
  cursor: pointer;
  height: 5rem;
  width: 100%;
  text-align: left;
  padding: 0 1rem;
}

.c2 {
  margin: 0 1rem;
  font-size: 1.4rem;
  width: 100%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-transition: -webkit-transform 0.3s linear;
  -webkit-transition: transform 0.3s linear;
  transition: transform 0.3s linear;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

<div
  class="c0"
>
  <button
    class="c1"
  >
    
    <p
      class="c2"
    >
      label 1
    </p>
    <span
      class="c3"
    >
      <svg
        aria-hidden="true"
        class="c4"
        focusable="false"
        size="20"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
          fill="#00222b"
        />
      </svg>
    </span>
  </button>
</div>
`;
